src/core/providers/ogr/qgsogrproviderutils.cpp

```diff
@@ -3106,6 +3106,92 @@ OGRErr QgsOgrLayer::GetExtent( OGREnvelope *psExtent, bool bForce )
   return OGR_L_GetExtent( hLayer, psExtent, bForce );
 }
 
+OGRErr QgsOgrLayer::GetExtent3D( OGREnvelope3D *psExtent3D, bool bForce )
+{
+  QMutexLocker locker( &ds->mutex );
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3,9,0)
+  return OGR_L_GetExtent3D( hLayer, /* iGeomField = */ 0, psExtent3D, bForce );
+#else
+
+  OGRErr err = OGRERR_UNSUPPORTED_OPERATION;
+  const char *geomCol = OGR_L_GetGeometryColumn( hLayer );
+  if ( geomCol != NULL && strlen( geomCol ) > 0 )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "WITH geomCol: %1" ).arg( geomCol ), 3 );
+
+    err = OGR_L_GetExtent( hLayer, psExtent3D, bForce );
+    if ( err == OGRERR_NONE )
+    {
+      err = OGRERR_UNSUPPORTED_OPERATION;
+      QByteArray geomColQuoted = QgsOgrProviderUtils::quotedIdentifier( geomCol, name() );
+      QByteArray sql = "SELECT MIN(ST_MinZ("
+                       + geomColQuoted
+                       + ")), MAX(ST_MaxZ("
+                       + geomColQuoted
+                       + ")) FROM "
+                       + QgsOgrProviderUtils::quotedIdentifier( name(), name() );
+      QgsDebugMsgLevel( QStringLiteral( "sql: %1" ).arg( sql.toStdString().c_str() ), 3 );
+
+      CPLPushErrorHandler( CPLQuietErrorHandler );
+      QgsOgrLayerUniquePtr l = ExecuteSQL( sql );
+      CPLPopErrorHandler();
+      if ( l )
+      {
+        gdal::ogr_feature_unique_ptr f( l->GetNextFeature() );
+        if ( f )
+        {
+          psExtent3D->MinZ = OGR_F_GetFieldAsDouble( f.get(), 0 );
+          psExtent3D->MaxZ = OGR_F_GetFieldAsDouble( f.get(), 1 );
+          QgsDebugMsgLevel( QStringLiteral( "done with Z! %1/%2" ).arg( psExtent3D->MinZ ).arg( psExtent3D->MaxZ ), 3 );
+          err = OGRERR_NONE;
+        }
+        ResetReading();
+      }
+    }
+  }
+
+  if ( err != OGRERR_NONE )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "Will computeExtent3DSlowly!" ), 3 );
+    err = computeExtent3DSlowly( psExtent3D );
```

Comment:
Don't we still have a performance regression here? 

Ie in the situation:

- a driver which doesn't support fast 3d extent retrieval from OGR_L_GetExtent3D OR on older GDAL and the SELECT ST_MinZ... fails. 
- dataset has z
- the driver DOES support FAST retrieval of 2 extent
- ONLY the api to retrieve the 2d extent is called (eg to zoom to layer extent on loading the layer)

In that situation we're now forcing a complete table traversal when asking the provider for the 2d extent, because the 3d extent is calculated first. Whereas before we could quickly retrieve the 2d extent without table traversal...

Reply:
hi @nyalldawson,

if the layer has z data and we have the 2D fast retrieval process for this driver, we still have to retrieve the minZ and maxZ data by using something like `SELECT ST_min(Z), ST_max(Z) FROM table`, which scan all rows. Also, by default for gdal 3.9, if the driver does not have an OGR_L_GetExtent3D implementation, gdal will scan all rows to compute the data (this is the same for the 2D extent).

Even if we use a second field to store the 2d extent when only the 2d extent is asked for a 3d layer, at some point, the 3D extent could be asked and we will have to scan all rows to compute it.

@rouault what do you think?

Reply:
> @rouault what do you think?

Will Extent3D will be called when adding a new vector layer ? If so, that might break some use cases. For example, currently you can add ``/vsizip//vsicurl/https://data.3dbag.nl/gpkg/v20231008/3dbag_nl.gpkg.zip`` (cf https://3dbag.nl/en/download . 19 GB as a ZIP) and use it reasonably with QGIS. But if Extent3D() is asked on it, then this will stall QGIS for hours.

Reply:
> Will Extent3D will be called when adding a new vector layer 

Yes it will, because `QgsOgrProvider::extent()` just calls `return extent3D().toRectangle();`, and the layer extent is always required when adding a layer to a project (eg to zoom to the first layer added's extent in a new project)

Reply:
Pragmatically I think we should leave QgsOgrProvider::extent() untouched, and ensure that we NEVER ask for the 3D extent in usual operations (ie not when adding layers or opening the layer properties windows).

Reply:
> Pragmatically I think we should leave QgsOgrProvider::extent() untouched, and ensure that we NEVER ask for the 3D extent in usual operations

I also concur with this

Reply:
I have updated the PR to:
* use the old extent 2D code
* use 2 different field members to save the 2D extent and the 3D extent

The behaviour is now what you expected: we only load 3D extent when the 3D extent is asked.

Furthermore we have found a strange behaviour: 
a geopackage which contains 3D data but according to the metadata, the geometry is 2D. For example, the `lod12_2D` layer from the `dbag_nl.gpkg.zip file.   
In that case, the 3D extent will never be computed because the provider reads the extent information from the metadata which says this is a 2D layer.

This behaviour is not induced by this PR but just "revealed".

Reply:
> I have updated the PR to:
> 
>     * use the old extent 2D code
> 
>     * use 2 different field members to save the 2D extent and the 3D extent

LGTM !

> Furthermore we have found a strange behaviour:
> a geopackage which contains 3D data but according to the metadata, the geometry is 2D. For example, the `lod12_2D` layer from the `dbag_nl.gpkg.zip file.
> In that case, the 3D extent will never be computed because the provider reads the extent information from the metadata which says this is a 2D layer.

Ok, that's interesting. That lod12_2D layer doesn't set the Z=1 or 2 hint in the gpkg_geometry_columns, which is used by the driver to infer the dimensionality of the layer geometry type. I've fixed the writing side of the driver to set Z=2 (may contain 3D geoms) when one is encountered in https://github.com/OSGeo/gdal/pull/9147 . That said, it is a bit weird that the data producer generates a layre called 2D but which contains only 3D geometries...

Reply:
> That said, it is a bit weird that the data producer generates a layre called 2D but which contains only 3D geometries...

indeed...